torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp

```diff
@@ -396,37 +407,116 @@ class DoubleBufferInserter : private kir::ExprMutator {
 
   void insert(
       kir::ForLoop* double_buffer_loop,
-      const std::vector<UnaryOp*>& loads) {
+      const std::vector<Expr*>& loads) {
     auto prologue_loop = DoubleBufferLoopCloner::clone(
         double_buffer_loop, loads, LoopType::Prologue);
     registerInsertBefore(double_buffer_loop, prologue_loop);
 
     auto write_to_smem =
-        std::any_of(loads.begin(), loads.end(), [](const UnaryOp* uop) {
-          return uop->out()->as<TensorView>()->getMemoryType() ==
+        std::any_of(loads.begin(), loads.end(), [](const Expr* expr) {
+          return expr->output(0)->as<TensorView>()->getMemoryType() ==
               MemoryType::Shared;
         });
 
     // RAW sync is not inserted for double buffered tensors. The only
     // exception is the prologue load.
+    bool insert_cpasync_wait = false;
     if (write_to_smem) {
       // Here the initial sync before entering double buffer loop is
-      //  inserted. It will also need to sync with async gmem loads.
+      //  inserted.
+
+      // If any of the double buffered tensor in this double buffer
+      //  loop is async copy. We want to wait for the gmem loads to
+      //  finish before synchronizing the block.
+      if (std::any_of(loads.begin(), loads.end(), ir_utils::isCpAsyncOp)) {
+        auto cp_async_wait = IrBuilder::create<kir::CpAsyncWait>();
+        registerInsertBefore(double_buffer_loop, cp_async_wait);
+        insert_cpasync_wait = true;
+      }
+
+      // Insert the initial block sync before entering main loop.
+      //
+      // TODO: we may not need this sync at all if we didn't need
+      //  to sync the block before. This information is in sync info.
+      //  In this case shared mem double buffering "collapses" into
+      //  register double buffering.
       auto sync = IrBuilder::create<kir::BlockSync>(false);
       registerInsertBefore(double_buffer_loop, sync);
     }
 
     auto main_loop = DoubleBufferLoopCloner::clone(
         double_buffer_loop, loads, LoopType::Main);
+
     registerReplace(double_buffer_loop, main_loop);
 
+    // Insert the wait instruction in this pass instead
+    //  of relying on WAR sync pass to do it.
+    // The WAR sync pass today would insert the wait function
+    //  exactly where we need it but the purpose of this wait
+    //  insertion isn't exactly WAR protection.
+    //
+    // TODO: [Double Buffer Sync]
```

Comment:
This is part of initial attempt to formalize double buffer syncs. 

We currently rely on WAR sync insertion pass to synchronize double buffered parallel read write. 

We could have just used WAR sync pass again to insert `cp.async`, but this is when we really coincide with the right answer but doing it the wrong way.

For quick reference, here is my zeroth-order summary of RAW and WAR hazard:

A RAW (read-after-write) sync/barrier is needed when we want to make sure a `write` finishes so we can start a dependent `read`.

A WAR (write-after-read) sync/barrier is needed when we want to make sure a `read` finishes so we can start an overlapping `write`.

In the case of double buffered access, below is the current way we generate code:

(case 1, more cases below)
```cpp

__shared__ T.
...  // prolog
for i in ...
  T[1-i%2] = ...
  ... = T[i%2]
  __sync()         //This sync is a WAR sync
```

This expr ordering can work with WAR sync, and actually RAW sync pass alone without WAR wouldn't be able to handle it correctly. Mechanically if we use WAR sync pass to insert a `cp.async.wait`, it'd look like:

```cpp

__shared__ T.
...  // prolog
for i in ...
  T[1-i%2] = ...
  ... = T[i%2]
  cp.async.wait // The wait inserted by WAR pass
  __sync()         //This sync is a WAR sync
```

This is correct code but wasn't particularly meaningful/robust because, 
 a cp.async.wait is an instruction that waits for a shared mem write to finish, and `back to the quick ref above, it shouldn't have been a WAR sync pass that wanted that inserted`, because a WAR sync makes sure a `read` finishes instead. 

But the reason this still works is that double buffer sync is more flexible on the insertion points and actually for `cp.async.wait` can be even more flexible because it is RAW only protection. But there are certain rules for double buffered sync as well so we should eventually model it as a separate family of barrier in follow ups. The rule for inserting the double buffer sync is:

`The producer and consumer of the double buffered tensor in the double buffer loop need to be on the same side of the sync`.

And we can currently use this rule for cp.async.wait for now as well, even though it can be more flexible.

In concrete examples, the following 3 patterns are valid as well as case 1 above:

case 2
```cpp

__shared__ T.
...  // prolog
for i in ...
  __sync()         //This sync is a RAW sync
  ... = T[i%2]
  T[1-i%2] = ...
```

case 3
```cpp

__shared__ T.
...  // prolog
for i in ...
  ... = T[i%2]
  T[1-i%2] = ...
  __sync()         //This is 
```

case 4
```cpp

__shared__ T.
...  // prolog
for i in ...
  __sync()         //This is again neither WAR or RAW
  T[1-i%2] = ...
  ... = T[i%2]
```

Reply:
What I'm currently trying to do is adding new sync logic to the double buffering pass, and in follow ups eventually we'd want to pull the double buffer logic out of the RAW/WAR insertion pass. 

Once we model this correctly we'd be able to re-order/interleave the expressions in double buffered loop, which in a few cases bring perf benefit.

Reply:
OK, let's leave this as a TODO.